src/dealer/DealerWallet.ts

```diff
@@ -0,0 +1,515 @@
+      // always be negative
+      btc: -netSize,
+
+      usd: -netSize * btcPrice,
+      estimatedLiquidationPrice,
+      collateralUsed, // USD
+      maintenanceMarginRequirement, // start at 0.03 but increase with position side
+
+      collateral, // in USD
+
+      totalAccountValue,
+
+      // if there is no collateral, marginFraction will be null. this is equivalent to infinite leverage.
+      leverage: marginFraction ? 1 / marginFraction : Number.POSITIVE_INFINITY,
+    }
+  }
+
+  static getExposureRatio({ usdLiability, usdExposure }) {
+    return {
+      ratio: usdExposure / usdLiability,
+      diff: usdExposure - usdLiability,
+    }
+  }
+
+  // we need to rebalance when price is increasing/decreasing.
+  // if price increase, then the short position risk being liquidated
+  // so we need to sned btc from the node to the exchange
+  // if price decrease, then there would be too much btc on the exchange
+  // the account won't be at the irsk of being liquidated in this case
+  // but then the custody risk of the exchange increases
+  static isRebalanceNeeded({ usdLiability, btcPrice, usdCollateral }) {
+    type IDepositOrWithdraw = "withdraw" | "deposit" | null
+
+    let usdAmountDiff
+    let depositOrWithdraw: IDepositOrWithdraw = null
+
+    if (!usdLiability) {
+      return { depositOrWithdraw }
+    }
+
+    // this is an approximation of leverage, because usdLiability is an internal value
+    // and not necessary exactly 100% of what is in the exchange
+    // ie: it should be bounded by
+    // const LOW_SAFEBOUND_RATIO_SHORTING = 0.98
+    // and
+    // const HIGH_SAFEBOUND_RATIO_SHORTING = 1
+    const leverage = usdLiability / usdCollateral
+
+    // under leveraged
+    // no imminent risk (beyond exchange custory risk)
+    if (leverage < yamlConfig.hedging.LOW_BOUND_LEVERAGE) {
+      const targetUsdCollateral = usdLiability / yamlConfig.hedging.LOW_SAFEBOUND_LEVERAGE
+      usdAmountDiff = usdCollateral - targetUsdCollateral
+      depositOrWithdraw = "withdraw"
+    }
+
+    // over leveraged
+    // our collateral could get liquidated if we don't rebalance
+    else if (leverage > yamlConfig.hedging.HIGH_BOUND_LEVERAGE) {
+      const targetUsdCollateral =
+        usdLiability / yamlConfig.hedging.HIGH_SAFEBOUND_LEVERAGE
+      usdAmountDiff = targetUsdCollateral - usdCollateral
+      depositOrWithdraw = "deposit"
+    }
+
+    const btcAmount = usdAmountDiff / btcPrice
+
+    if (depositOrWithdraw) {
+      assert(btcAmount > 0)
+      // amount more than 3x of collateral should not happen
+      // although it could happen the first time the dealer is launched?
+      assert(btcAmount < 3 * usdLiability * btcPrice)
+    }
+
+    return { btcAmount, depositOrWithdraw }
+  }
+
+  static isOrderNeeded({ usdLiability, usdExposure, btcPrice }) {
+    const { ratio } = this.getExposureRatio({ usdLiability, usdExposure })
+
+    let usdOrderAmount
+    let buyOrSell: IBuyOrSell = null
```

Comment:
Here again... please use a more descriptive value for `uninitialised` than `null`.